proposals/20200224-combined-logging.md

```diff
@@ -0,0 +1,72 @@
+# Combined Logging
+
+Support the ability to view all of the container logs for a given workload. Currently you must request each container
+log separately.
+
+## Goals
+ - Combine the logging output from multiple containers.
+ - Tag/mark which log lines are from which container.
+ - Provide a list of which container logs are included in the stream.
+ - Only send content when there are new log entries.
+
+## Non-Goals
+ - Server-side filtering
+
+## Log Streaming
+
+### Current implementation
+Our existing log "streaming" is done via a log endpoint with a JSON request. We explicitly register a container logs
+handler in `internal/api/api.go`.
+
+    s.HandleFunc("/logs/namespace/{namespace}/pod/{pod}/container/{container}", containerLogsHandler(ctx, a.dashConfig))
+
+When visiting the Logs tab we generate an XHR matching the URL pattern we've registered for container logs. This XHR
+request is generated using the selected container from the listed of containers. This list of containers is provided
+by Octant to the Log component. A user can only view a single containers logs at a time.
+
+Example request:
+
+    http://localhost:7777/api/v1/logs/namespace/octant-demo/pod/kuard-7bbd76c779-6n2p4/container/kuard
+
+Octant responds with a JSON object that contains the entirety of the log data each request. The request is repeated via
+a polling mechanism in the PodLogsStreamer, every 5 second. When a new container is selected from the dropdown, the
+existing poll is canceled and a new one is created with a request that uses the newly selected container.
+
+### Suggested implementation
+Add a sentinel value of `AllContainers` to Octant and make this the default choice when requesting logs.
+
+Add a set of new actions to Octant to handle starting a log stream.
+
+Add a new EventType to Octant for log streams that the WebsocketService can subscribe to.
+
+Extend our existing log entry to include the container the log entry is from.
+
+#### Stream Aggregation
+Create a go routine that wraps the `Stream` call for each containers Octant will stream logs for.
```

Comment:
Is this a good opportunity to create a better logging interface? I’m not 100% sure what it should look like but the ideas seems to be in the right direction.

Reply:
Yeah, I think we will ultimately end up with a different interface. I will walk through some ideas and update the proposal to reflect the new design.

I'm thinking this change will create an `internal/logging` module which will house the action handlers, streaming functions, and helpers.